### PR TITLE
[ROCM] Support TF_ROCM_CLANG for builds with clang host compiler

### PR DESCRIPTION
**Pull request by draganmladjenovic**
Remove TF_NEED_CLANG workarounds

### CODE REVIEW

**Comment by i-chaochen** [comment]
retest Ubuntu-CPU please

**Comment by jayfurmanek** [comment]
We'll also need to update the symlink of rocm.bazelrc to point to gpu.bazelrc (instead of gpu_gcc.bazelrc)

**Comment by jayfurmanek** [comment]
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/rocm.bazelrc

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-GPU-single please
Retest Ubuntu-CPU please

**Comment by i-chaochen** [comment]
I wonder what's situation for this PR? are we capable of switching to clang now?

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-CPU please.

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-GPU-multi please.

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-CPU please.
Retest Ubuntu-GPU-single please.

**Comment by draganmladjenovic** [comment]
@jayfurmanek I think this one is ready. I slightly dislike having to change root .bazelrc in the last commit. Alternative is to create /userlocal/ in Dockerfile.rocm and point bazel in run_gpu_single/multi.sh to use --confing=/userlocal/gpu.bazelrc.

**Comment by draganmladjenovic** [comment]
@i-chaochen  Where can I update the docker image for ThirdParty-XLA ci?

**Comment by i-chaochen** [comment]
ThirdParty-XLA CI is using the same tensorflow docker image (```rocm/tensorflow-build:latest-python3.9-rocm5.7.0```), may I ask what do you mean you want to update it, and why you want to update it?

**Comment by draganmladjenovic** [comment]
> ThirdParty-XLA CI is using the same tensorflow docker image (`rocm/tensorflow-build:latest-python3.9-rocm5.7.0`), may I ask what do you mean you want to update it, and why you want to update it?

It is missing clang-16, so the PR fails, and cannot be merged.

**Comment by i-chaochen** [comment]
> rocm/tensorflow-build:latest-python3.9-rocm5.7.0

I think you need to add clang-16 into these three:

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm 

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.rt 

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.ub20
 

```rocm/tensorflow-build:latest-python3.9-rocm5.7.0``` will be updated the latest build.

**Comment by draganmladjenovic** [comment]
> > rocm/tensorflow-build:latest-python3.9-rocm5.7.0
> 
> I think you need to add clang-16 into these three:
> 
> https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm
> 
> https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.rt
> 
> https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.ub20
> 
> `rocm/tensorflow-build:latest-python3.9-rocm5.7.0` will be updated the latest build.
Ok. I'll split this PR.

**Comment by draganmladjenovic** [comment]
> > ThirdParty-XLA CI is using the same tensorflow docker image (`rocm/tensorflow-build:latest-python3.9-rocm5.7.0`), may I ask what do you mean you want to update it, and why you want to update it?
> 
> It is missing clang-16, so the PR fails, and cannot be merged.

My bad. This CI is not blocking the merge. @jayfurmanek The only thing that is missing is the review. Should I include more people into reviewers?

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc
Comment (i-chaochen) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc:20]: Since we're switching to use gpu.bazelrc, could you also add [xla_cpp_filters](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu_gcc.bazelrc#L184-L187) from gpu_gcc.bazelrc ? Thanks!

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc
Comment (jayfurmanek) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc]: This line pulls in the cuda config from the .bazelrc.
We'll need a "release_rocm_linux" config there I think..

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc
Comment (jayfurmanek) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc]: Can we update to clang-17 to match what's upstream now?

Comment (jayfurmanek) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc]: That will need to be changed in the container as well (merged separately)

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc
Comment (jayfurmanek) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc]: I think we want to use ldd for the linker

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc
Comment (draganmladjenovic) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc]: > Can we update to clang-17 to match what's upstream now?

Thanks. Missed that. Will update.

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc
Comment (draganmladjenovic) [reply tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc]: I think I had some issues getting lld on all docker images. I will have to check that.

**Comment by jayfurmanek** [comment]
retest gpu-pycpp please

**Comment by jayfurmanek** [comment]
retest gpu-non-pip-multi please

**Comment by jayfurmanek** [comment]
retest Ubuntu-GPU-multi please

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-GPU-single please.

**Review by jayfurmanek — APPROVED** [review]
Looks good!
Please tag the repo before and after merging:
pre-clang-merge
post-clang-merge

**Pull request merged**
